File: go-client/internal/storage/vault.go
// Package storage provides encrypted vault and file store management.
package storage

import (
	"crypto/ed25519"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/cisc468/p2p-project/internal/crypto"
)

// Vault is the in-memory representation of the decrypted vault.
type Vault struct {
	DataDir    string
	VaultPath  string
	VaultKey   []byte
	IdentitySK ed25519.PrivateKey
	IdentityPK ed25519.PublicKey
	Contacts   []Contact
	Settings   Settings
	CreatedAt  int64
}

// Contact represents a known peer in the contact book.
type Contact struct {
	PeerID           string   `json:"peer_id"`
	Alias            string   `json:"alias"`
	PublicKey        string   `json:"public_key"` // base64url of raw 32-byte Ed25519 pubkey
	AddedAt          int64    `json:"added_at"`
	LastSeen         int64    `json:"last_seen"`
	MigrationHistory []string `json:"migration_history"`
}

// Settings holds user preferences.
type Settings struct {
	DisplayName              string `json:"display_name"`
	RequireConsentForTransfer bool   `json:"require_consent_for_transfer"`
}

// vaultPlaintext is the JSON structure inside the encrypted vault.
type vaultPlaintext struct {
	Identity struct {
		PrivateKey string `json:"private_key"`
		PublicKey  string `json:"public_key"`
		CreatedAt  int64  `json:"created_at"`
	} `json:"identity"`
	Contacts []Contact `json:"contacts"`
	Settings Settings  `json:"settings"`
}

// vaultOuter is the on-disk vault.json structure.
type vaultOuter struct {
	VaultVersion int `json:"vault_version"`
	KDF          string `json:"kdf"`
	KDFParams    struct {
		Time        int    `json:"time"`
		Memory      int    `json:"memory"`
		Parallelism int    `json:"parallelism"`
		Salt        string `json:"salt"`
		KeyLen      int    `json:"keylen"`
	} `json:"kdf_params"`
	Ciphertext string `json:"ciphertext"`
}

// NewVault creates a new Vault handle for the given data directory.
func NewVault(dataDir string) *Vault {
	return &Vault{
		DataDir:   dataDir,
		VaultPath: filepath.Join(dataDir, "vault.json"),
	}
}

// PeerID returns the base64url peer ID of this vault's identity.
func (v *Vault) PeerID() string {
	return crypto.PeerIDFromPublicKey(v.IdentityPK)
}

// Exists checks if vault.json exists on disk.
func (v *Vault) Exists() bool {
	_, err := os.Stat(v.VaultPath)
	return err == nil
}

// Create initializes a new vault with a fresh identity keypair.
func (v *Vault) Create(password, displayName string) error {
	os.MkdirAll(filepath.Join(v.DataDir, "files"), 0700)
	os.MkdirAll(filepath.Join(v.DataDir, "migrations"), 0700)

	sk, pk, err := crypto.GenerateIdentityKeypair()
	if err != nil {
		return err
	}
	v.IdentitySK = sk
	v.IdentityPK = pk
	v.CreatedAt = time.Now().UnixMilli()
	v.Contacts = []Contact{}
	v.Settings = Settings{
		DisplayName:              displayName,
		RequireConsentForTransfer: true,
	}

	salt, err := crypto.GenerateSalt()
	if err != nil {
		return err
	}
	v.VaultKey = crypto.DeriveVaultKey(password, salt)
	return v.save(salt)
}

// Open decrypts and loads an existing vault.
func (v *Vault) Open(password string) error {
	if !v.Exists() {
		return errors.New("vault file not found — run 'init' first")
	}

	data, err := os.ReadFile(v.VaultPath)
	if err != nil {
		return err
	}
	var outer vaultOuter
	if err := json.Unmarshal(data, &outer); err != nil {
		return fmt.Errorf("parse vault: %w", err)
	}

	salt, err := crypto.B64URLDecode(outer.KDFParams.Salt)
	if err != nil {
		return err
	}
	v.VaultKey = crypto.DeriveVaultKey(password, salt)

	ciphertext, err := crypto.B64URLDecode(outer.Ciphertext)
	if err != nil {
		return err
	}
	plaintext, err := crypto.DecryptVault(v.VaultKey, ciphertext)
	if err != nil {
		return errors.New("incorrect password or corrupted vault file")
	}

	var vp vaultPlaintext
	if err := json.Unmarshal(plaintext, &vp); err != nil {
		return fmt.Errorf("parse vault plaintext: %w", err)
	}

	seed, err := crypto.B64URLDecode(vp.Identity.PrivateKey)
	if err != nil {
		return err
	}
	v.IdentitySK, err = crypto.SeedToPrivateKey(seed)
	if err != nil {
		return err
	}

	pkBytes, err := crypto.B64URLDecode(vp.Identity.PublicKey)
	if err != nil {
		return err
	}
	v.IdentityPK, err = crypto.BytesToPublicKey(pkBytes)
	if err != nil {
		return err
	}

	v.CreatedAt = vp.Identity.CreatedAt
	v.Contacts = vp.Contacts
	v.Settings = vp.Settings
	return nil
}

// Save re-encrypts and saves the vault.
func (v *Vault) Save() error {
	if v.VaultKey == nil {
		return errors.New("vault not open")
	}
	data, err := os.ReadFile(v.VaultPath)
	if err != nil {
		return err
	}
	var outer vaultOuter
	if err := json.Unmarshal(data, &outer); err != nil {
		return err
	}
	salt, err := crypto.B64URLDecode(outer.KDFParams.Salt)
	if err != nil {
		return err
	}
	return v.save(salt)
}

func (v *Vault) save(salt []byte) error {
	vp := vaultPlaintext{
		Contacts: v.Contacts,
		Settings: v.Settings,
	}
	vp.Identity.PrivateKey = crypto.B64URLEncode(crypto.PrivateKeyToSeed(v.IdentitySK))
	vp.Identity.PublicKey = crypto.B64URLEncode(crypto.PublicKeyToBytes(v.IdentityPK))
	vp.Identity.CreatedAt = v.CreatedAt

	plaintext, err := json.Marshal(vp)
	if err != nil {
		return err
	}
	encrypted, err := crypto.EncryptVault(v.VaultKey, plaintext)
	if err != nil {
		return err
	}

	outer := vaultOuter{
		VaultVersion: 1,
		KDF:          "argon2id",
		Ciphertext:   crypto.B64URLEncode(encrypted),
	}
	outer.KDFParams.Time = crypto.Argon2Time
	outer.KDFParams.Memory = crypto.Argon2Memory
	outer.KDFParams.Parallelism = crypto.Argon2Parallelism
	outer.KDFParams.Salt = crypto.B64URLEncode(salt)
	outer.KDFParams.KeyLen = crypto.Argon2KeyLen

	data, err := json.MarshalIndent(outer, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(v.VaultPath, data, 0600)
}

// AddContact adds or updates a contact.
func (v *Vault) AddContact(peerID string, publicKeyBytes []byte, alias string) {
	now := time.Now().UnixMilli()
	for i, c := range v.Contacts {
		if c.PeerID == peerID {
			v.Contacts[i].LastSeen = now
			return
		}
	}
	if alias == "" {
		alias = peerID[:8]
	}
	v.Contacts = append(v.Contacts, Contact{
		PeerID:           peerID,
		Alias:            alias,
		PublicKey:         crypto.B64URLEncode(publicKeyBytes),
		AddedAt:          now,
		LastSeen:         now,
		MigrationHistory: []string{},
	})
}

// GetContact looks up a contact by peer_id.
func (v *Vault) GetContact(peerID string) *Contact {
	for i, c := range v.Contacts {
		if c.PeerID == peerID {
			return &v.Contacts[i]
		}
	}
	return nil
}

// UpdateContactKey updates a contact's key after migration.
func (v *Vault) UpdateContactKey(oldPeerID, newPeerID string, newPublicKeyBytes []byte) bool {
	for i, c := range v.Contacts {
		if c.PeerID == oldPeerID {
			v.Contacts[i].MigrationHistory = append(v.Contacts[i].MigrationHistory, oldPeerID)
			v.Contacts[i].PeerID = newPeerID
			v.Contacts[i].PublicKey = crypto.B64URLEncode(newPublicKeyBytes)
			v.Contacts[i].LastSeen = time.Now().UnixMilli()
			return true
		}
	}
	return false
}

// GetDisplayName returns the configured display name.
func (v *Vault) GetDisplayName() string {
	if v.Settings.DisplayName == "" {
		return "peer"
	}
	return v.Settings.DisplayName
}

// ChangePassword re-encrypts the vault with a new password.
func (v *Vault) ChangePassword(newPassword string) error {
	if v.VaultKey == nil {
		return errors.New("vault not open")
	}
	salt, err := crypto.GenerateSalt()
	if err != nil {
		return err
	}
	v.VaultKey = crypto.DeriveVaultKey(newPassword, salt)
	return v.save(salt)
}

File: python-client/p2pshare/storage/vault.py
"""Encrypted vault for identity key, contacts, and settings.

The vault is stored as vault.json in the data directory.
It is encrypted with AES-256-GCM using a key derived from the user's password via Argon2id.
"""

import json
import os
import time

from p2pshare.crypto.keys import (
    generate_identity_keypair,
    private_key_to_seed,
    seed_to_private_key,
    public_key_to_bytes,
    bytes_to_public_key,
    b64url_encode,
    b64url_decode,
    peer_id_from_public_key,
)
from p2pshare.crypto.storage import (
    derive_vault_key,
    generate_salt,
    encrypt_vault,
    decrypt_vault,
    ARGON2_TIME,
    ARGON2_MEMORY,
    ARGON2_PARALLELISM,
    ARGON2_KEYLEN,
)
from p2pshare.errors import VaultError


class Vault:
    """In-memory representation of the decrypted vault."""

    def __init__(self, data_dir: str):
        self.data_dir = data_dir
        self.vault_path = os.path.join(data_dir, "vault.json")
        self.vault_key: bytes | None = None
        self.identity_sk = None
        self.identity_pk = None
        self.contacts: list[dict] = []
        self.settings: dict = {}
        self._created_at: int = 0

    @property
    def peer_id(self) -> str:
        return peer_id_from_public_key(self.identity_pk)

    def exists(self) -> bool:
        return os.path.exists(self.vault_path)

    def create(self, password: str, display_name: str = "peer") -> None:
        """Create a new vault with a fresh identity keypair."""
        os.makedirs(self.data_dir, exist_ok=True)
        os.makedirs(os.path.join(self.data_dir, "files"), exist_ok=True)
        os.makedirs(os.path.join(self.data_dir, "migrations"), exist_ok=True)

        self.identity_sk, self.identity_pk = generate_identity_keypair()
        self._created_at = int(time.time() * 1000)
        self.contacts = []
        self.settings = {
            "display_name": display_name,
            "require_consent_for_transfer": True,
        }

        salt = generate_salt()
        self.vault_key = derive_vault_key(password, salt)
        self._save(salt)

    def open(self, password: str) -> None:
        """Decrypt and load an existing vault."""
        if not self.exists():
            raise VaultError("Vault file not found. Run 'init' first.")

        with open(self.vault_path, "r") as f:
            outer = json.load(f)

        salt = b64url_decode(outer["kdf_params"]["salt"])
        self.vault_key = derive_vault_key(password, salt)

        payload = b64url_decode(outer["ciphertext"])
        try:
            plaintext = decrypt_vault(self.vault_key, payload)
        except Exception:
            raise VaultError("Incorrect password or corrupted vault file.")

        data = json.loads(plaintext.decode("utf-8"))
        self.identity_sk = seed_to_private_key(b64url_decode(data["identity"]["private_key"]))
        self.identity_pk = bytes_to_public_key(b64url_decode(data["identity"]["public_key"]))
        self._created_at = data["identity"]["created_at"]
        self.contacts = data.get("contacts", [])
        self.settings = data.get("settings", {})

    def save(self) -> None:
        """Re-encrypt and save the vault with the current vault key."""
        if self.vault_key is None:
            raise VaultError("Vault not open")
        # Read existing salt
        with open(self.vault_path, "r") as f:
            outer = json.load(f)
        salt = b64url_decode(outer["kdf_params"]["salt"])
        self._save(salt)

    def _save(self, salt: bytes) -> None:
        """Encrypt and write vault.json."""
        plaintext_data = {
            "identity": {
                "private_key": b64url_encode(private_key_to_seed(self.identity_sk)),
                "public_key": b64url_encode(public_key_to_bytes(self.identity_pk)),
                "created_at": self._created_at,
            },
            "contacts": self.contacts,
            "settings": self.settings,
        }
        plaintext = json.dumps(plaintext_data, separators=(",", ":")).encode("utf-8")
        encrypted = encrypt_vault(self.vault_key, plaintext)

        outer = {
            "vault_version": 1,
            "kdf": "argon2id",
            "kdf_params": {
                "time": ARGON2_TIME,
                "memory": ARGON2_MEMORY,
                "parallelism": ARGON2_PARALLELISM,
                "salt": b64url_encode(salt),
                "keylen": ARGON2_KEYLEN,
            },
            "ciphertext": b64url_encode(encrypted),
        }

        with open(self.vault_path, "w") as f:
            json.dump(outer, f, indent=2)

    def add_contact(self, peer_id: str, public_key_bytes: bytes, alias: str = "") -> None:
        """Add a new contact to the vault."""
        for c in self.contacts:
            if c["peer_id"] == peer_id:
                c["last_seen"] = int(time.time() * 1000)
                return
        self.contacts.append({
            "peer_id": peer_id,
            "alias": alias or peer_id[:8],
            "public_key": b64url_encode(public_key_bytes),
            "added_at": int(time.time() * 1000),
            "last_seen": int(time.time() * 1000),
            "migration_history": [],
        })

    def get_contact(self, peer_id: str) -> dict | None:
        """Look up a contact by peer_id."""
        for c in self.contacts:
            if c["peer_id"] == peer_id:
                return c
        return None

    def update_contact_key(self, old_peer_id: str, new_peer_id: str, new_public_key_bytes: bytes) -> bool:
        """Update a contact's key after key migration. Returns True if found and updated."""
        for c in self.contacts:
            if c["peer_id"] == old_peer_id:
                c["migration_history"].append(old_peer_id)
                c["peer_id"] = new_peer_id
                c["public_key"] = b64url_encode(new_public_key_bytes)
                c["last_seen"] = int(time.time() * 1000)
                return True
        return False

    def get_display_name(self) -> str:
        return self.settings.get("display_name", "peer")

    def change_password(self, new_password: str) -> None:
        """Re-encrypt the vault with a new password."""
        if self.vault_key is None:
            raise VaultError("Vault not open")
        salt = generate_salt()
        self.vault_key = derive_vault_key(new_password, salt)
        self._save(salt)
